Update the README to align with the documents (#243)
android/user-interface-samples: AppWidget/README.md

List of API changes
------------

> # [Improve your app's widget picker experience](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#improve-widget-picker-experience)

Reviewer: ```suggestion
# [Improve your app's widget picker experience](https://developer.android.com/about/versions/12/features/widgets#improve-widget-picker-experience)
```

Author: Done

> ## [Add scalable widget previews to the widget picker](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#add-scalable-widget-previews)

Reviewer: ```suggestion
## [Add scalable widget previews to the widget picker](https://developer.android.com/about/versions/12/features/widgets#add-scalable-widget-previews)
```

Author: Done.

  In Android 12, the widget preview displayed in the widget picker consists of a scalable preview, which you’ll provide as an XML layout set to the widget's default size. Previously, the widget preview was a static drawable resource, in some cases leading to previews not accurately reflecting widgets after they were added to the home screen.

>   To implement scalable widget previews, use the [previewLayout](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#previewLayout) attribute of the appwidget-provider element to provide an XML layout instead:

Reviewer 2: `previewLayout`

  [`app_widget_info_checkbox_list`](app/src/main/res/xml/app_widget_info_checkbox_list.xml\#L50)
  ```xml

>     <appwidget-provider

Reviewer: Nit: why is this indented?

Author: It wasn't intended. Removed the indentation

        android:previewLayout="@layout/widget_grocery_list"
    ... />
  ```

> ## [Add a description for your widget](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#add-widget-description)

Reviewer: ```suggestion
## [Add a description for your widget](https://developer.android.com/about/versions/12/features/widgets#add-widget-description)
```

Author: Done.

  In Android 12, you can optionally provide a description for the widget picker to display for your widget. 

>   Provide a description for your widget using the description attribute of appwidget-provider:

Reviewer 2: `appwidget-provider`

  [`app_widget_info_checkbox_list`](app/src/main/res/xml/app_widget_info_checkbox_list.xml\#L40)
  ```xml

>   <appwidget-provider

Reviewer: Ditto, why indented?

Author: It wasn't intended. Removed the indentation.

      android:description="@string/app_widget_grocery_list_description"
      ... />
  ```
   
  <img src="screenshots/widget_preview.png"
    alt="Image of the widget with description and previewLayout"

>     title="Image of the widget with description and previewLayout"

Reviewer 2: `previewLayout`

    />

> # [Apply dynamic colors](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#dynamic-colors)

Reviewer: ```suggestion
# [Apply dynamic colors](https://developer.android.com/about/versions/12/features/widgets#dynamic-colors)
```

Author: Done.

  In Android 12, a widget can use the device theme colors for buttons, backgrounds, and other components. This enables smoother transitions and consistency across different widgets.

  <img src="screenshots/dynamic_coloring_light_theme.png" width="240px"
      alt="screenshot for dynamic coloring on light theme"
      title="screenshot for dynamic coloring on light theme" />
  <img src="screenshots/dynamic_coloring_dark_theme.png" width="240px"
      alt="screenshot for dynamic coloring on dark theme"
      title="screenshot for dynamic coloring on dark theme" />

> ## [Backward-compatibility with dynamic colors](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#backward-compatibility_with_dynamic_colors)

Reviewer: ```suggestion
## [Backward-compatibility with dynamic colors](https://developer.android.com/about/versions/12/features/widgets#backward-compatibility_with_dynamic_colors)
```

Author: Done.

>   We recommend creating a custom theme and overriding it when running on devices with Android 12+. The following examples show how to do this with various types of XML files:

Reviewer 2: ... when running on devices with Android 12 or later.

>   We recommend creating a custom theme and overriding it for Android 12. The following examples show how to do this with various types of XML files:

Reviewer: ```suggestion
  We recommend creating a custom theme and overriding it when running on devices with Android 12+. The following examples show how to do this with various types of XML files:
```

Author: Done.

  </LinearLayout>
  ```

> # [Implement rounded corners](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#rounded-corner)

Reviewer: ```suggestion
# [Implement rounded corners](https://developer.android.com/about/versions/12/features/widgets#rounded-corner)
```

Author: Done.

  Android 12 introduces the following system parameters to set the radii of your widget's rounded corners:

  - `system_app_widget_background_radius`: The corner radius of the widget background, which will never be larger than 28dp.
  - `system_app_widget_inner_radius`: The corner radius of any view inside the widget. This is exactly 8dp less than the background radius to align nicely when using an 8dp padding.

> ## [Backward-compatibility with rounded corners](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#backward-compatibility_with_rounded_corners)

Reviewer: ```suggestion
## [Backward-compatibility with rounded corners](https://developer.android.com/about/versions/12/features/widgets#backward-compatibility_with_rounded_corners)
```

Author: Done.

      alt="screenshot for a widget with rounded corners"
      title="screenshot for a widget with rounded corners" />

> # [Make it easier to personalize widgets](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#personalize-widgets)

Reviewer: ```suggestion
# [Make it easier to personalize widgets](https://developer.android.com/about/versions/12/features/widgets#personalize-widgets)
```

Author: Done.

>   If you specify a configuration activity with the [configure](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#configure) attribute of [appwidget-provider](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo), the App Widget host launches that activity immediately after a user adds the widget to their home screen.

Reviewer 2: Enclose configure and appwidget-provider in backticks

  Android 12 adds new options to let you provide a better configuration experience for users.

> ## [Enable users to reconfigure placed widgets](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#reconfigure-widgets)

Reviewer: ```suggestion
## [Enable users to reconfigure placed widgets](https://developer.android.com/about/versions/12/features/widgets#reconfigure-widgets)
```

Author: Done.

>   To configure widgets that are labeled as reconfigurable, users can long-press the widget. This displays a Reconfigure button, which they can tap to change settings.

Reviewer 2: Boldface Reconfigure

  <img src="screenshots/widget_reconfigure.png"
      alt="screenshot for a widget with reconfigure button"
      title="screenshot for a widget with reconfigure button" />

>   Specify the `reconfigurable` flag in the `widgetFeatures` attribute of appwidget-provider:

Reviewer 2: `appwidget-provider`

  [`xml/app_widget_info_checkbox_list.xml`](app/src/main/res/xml/app_widget_info_checkbox_list.xml\#L39)
  ```xml
  <appwidget-provider
      android:configure="com.example.android.appwidget.ListWidgetConfigureActivity"
      android:widgetFeatures="reconfigurable"
      ... />
  ```
  

> ## [Use the widget's default configuration](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#use-default)

Reviewer: ```suggestion
## [Use the widget's default configuration](https://developer.android.com/about/versions/12/features/widgets#use-default)
```

Author: Done.

      ... />
  ```

# [Enable smoother transitions](https://developer.android.com/about/versions/12/features/widgets#enable-smoother-transitions)

  In Android 12, launchers provide a smoother transition when a user launches your app from a widget.

>   To enable this improved transition, use [android:id/background](https://developer.android.com/reference/android/R.id#background)  or android.R.id.background to identify your background element:

Reviewer 2: `android:id/background`

`android.R.id.background`

  ```xml
  <LinearLayout
    android:id="@android:id/background">
    ...
  </LinearLayout>

  ```
  <img src="screenshots/widget_smooth_transition.gif"
      alt="screen record of a smooth transition from the widget to the app"
      title="screen record of a smooth transition from the widget to the app" />

# [Add new compound buttons](https://developer.android.com/about/versions/12/features/widgets#new-compound-buttons)

  Android 12 adds new support for stateful behavior using the following existing components:

  - [Checkbox](https://developer.android.com/reference/kotlin/android/widget/CheckBox)
  - [Switch](https://developer.android.com/reference/kotlin/android/widget/Switch)

>   - [RadioButton](https://developer.android.com/reference/kotlin/android/widget/RadioButton)

Reviewer 2: `Checkbox`
`Switch`
`RadioButton`

  ```

>   See the documents for [fromPendingIntent](https://developer.android.com/reference/android/widget/RemoteViews.RemoteResponse#fromPendingIntent(android.app.PendingIntent)) and [fromFillInIntent](https://developer.android.com/reference/android/widget/RemoteViews.RemoteResponse#fromFillInIntent(android.content.Intent)) for more details.

Reviewer 2: `fromPendingIntent`

`fromFillInIntent`

# [Use simplified RemoteViews collections](https://developer.android.com/about/versions/12/features/widgets#leverage-simplified-remoteview-collections)

>   Android 12 adds the `setRemoteAdapter(int viewId, RemoteViews.RemoteCollectionItems items)` method, which lets your app pass along a collection directly when populating a ListView. Previously, when using a ListView, it was necessary to implement and declare a [RemoteViewService](https://developer.android.com/reference/android/widget/RemoteViewsService) to return 

Reviewer 2: `ListView`

  [RemoteViewsFactory](https://developer.android.com/reference/android/widget/RemoteViewsService.RemoteViewsFactory).

  If the collection doesn’t use a constant set of layouts (in other words, if some items are only sometimes present), use `setViewTypeCount` to specify the maximum number of unique layouts the collection can contain.

>   Here’s an example of how to implement simplified RemoteViews collections.

Reviewer 2: `RemoteViews`

  )
  ```

> # [Use improved APIs for widget sizes and layouts](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#refine-widget-sizes)

Reviewer: ```suggestion
# [Use improved APIs for widget sizes and layouts](https://developer.android.com/about/versions/12/features/widgets#refine-widget-sizes)
```

Author: Done.

  Starting in Android 12, you can provide more refined size attributes and more flexible layouts and by doing the following:

  1. [Specify additional widget sizing constraints](#specify-additional-widget-sizing-constraints)

>   2. [Provide responsive layouts](#provide-responsive-layouts) or [exact layouts](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#provide-exact-layouts)

Reviewer: ```suggestion
  2. [Provide responsive layouts](#provide-responsive-layouts) or [exact layouts](https://developer.android.com/about/versions/12/features/widgets#provide-exact-layouts)
```

Author: Done.

> ## [Specify additional widget sizing constraints](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#specify-widget-size-constraints)

Reviewer: ```suggestion
## [Specify additional widget sizing constraints](https://developer.android.com/about/versions/12/features/widgets#specify-widget-size-constraints)
```

Author: Done.

  Android 12 adds new APIs allowing you to ensure your widget is sized more reliably across different devices with varying screen sizes.

>   In addition to the existing [minWidth](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#minWidth), [minHeight](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#minHeight), [minResizeWidth](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#minResizeWidth), and [minResizeHeight](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#minResizeHeight) attributes, use the following new appwidget-provider attributes:

Reviewer 2: Put backticks around all APIs in this paragraph.

  - [targetCellWidth](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#targetCellWidth) and [targetCellHeight](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#targetCellHeight): Defines the target size of the widget in terms of launcher grid cells. If defined, these attributes are used instead of minWidth or minHeight. 

>   - [maxResizeWidth](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#maxResizeWidth) and [maxResizeHeight](https://developer.android.com/reference/android/appwidget/AppWidgetProviderInfo#maxResizeHeight): Defines the maximum size the launcher allows the user to resize the widget.

Reviewer 2: `maxResizeWidth`

       alt="Screenshot of a widget being resized"
       title="Screenshot of a widget being resized" />

> ## [Provide responsive layouts](https://developer.android.com/about/versions/12/features/widgets?db=dac-team#provide-responsive-layouts)

Reviewer: ```suggestion
## [Provide responsive layouts](https://developer.android.com/about/versions/12/features/widgets#provide-responsive-layouts)
```

Author: Done.
